Merging two pages about client libraries

JanStaschulat opened the pull request:
- inserted text from `About-Client-Interfaces.rst`in `About-ROS-2-Client-Libraries.rst`
- replaced C++11 by C++17
- see related PR: 
  - https://github.com/ros2/ros2_documentation/pull/2886
  - https://github.com/ros2/ros2_documentation/pull/2905

JanStaschulat committed:
copied content of 'About-Client-Interfaces' to 'About-ROS-2-Client-Libraries' and removed 'About-Client-Interfaces.rst'.

Signed-off-by: Jan Staschulat (CR/ADA1.2) <jan.staschulat@de.bosch.com>

JanStaschulat committed:
|GitHub|_ created verification error (see https://github.com/ros2/ros2_documentation/runs/7695375141?check_suite_focus=true), resolved this by using simple text 'GitHub'.

Signed-off-by: Jan Staschulat (CR/ADA1.2) <jan.staschulat@de.bosch.com>

clalancette commented on source/Concepts/About-ROS-2-Client-Libraries.rst:
```diff
@@ -32,6 +32,37 @@ Supported client libraries
+The ``rclcpp`` repository is located on GitHub at `ros2/rclcpp <https://github.com/ros2/rclcpp>`_ and contains the |package| ``rclcpp``.
+The generated |API| documentation is here:
+
+`api/rclcpp/index.html <http://docs.ros2.org/foxy/api/rclcpp/index.html>`_
```
So for Rolling, this is no longer the right URL.  We may as well fix it up:

```suggestion
`api/rclcpp/index.html <http://docs.ros.org/en/{DISTRO}/p/rclcpp>`_
```

clalancette replied:
(we unfortunately don't have this for rclpy at the moment, so the rclpy one can stay as-is)

clalancette commented on source/Concepts/About-ROS-2-Client-Libraries.rst:
```diff
@@ -32,6 +32,37 @@ Supported client libraries
+The ``rclpy`` repository is located on GitHub at `ros2/rclpy <https://github.com/ros2/rclpy>`_ and contains the |package| ``rclpy``.
+The generated |API| documentation is here:
+
+
```
Nit: remove one of these blank lines.

JanStaschulat replied:
resolved in [858d0da](https://github.com/ros2/ros2_documentation/pull/2919/commits/858d0daed4d54bbca30e0bdc2ac78c4add178d66).

clalancette reviewed (changes requested):
Besides the small change inline, we should also add in a redirect from the old (deleted) page to the new one.  So at the top of `About-ROS-2-Client-Libraries`, add in:

```
.. redirect-from::

   Concepts/About-Client-Interfaces
```

(see https://github.com/ros2/ros2_documentation/blob/rolling/source/Concepts/About-Logging.rst?plain=1 for an example)

JanStaschulat committed:
Update source/Concepts/About-ROS-2-Client-Libraries.rst

Co-authored-by: Chris Lalancette <clalancette@gmail.com>

JanStaschulat committed:
resolved reviewer comments

Signed-off-by: Jan Staschulat (CR/ADA1.2) <jan.staschulat@de.bosch.com>

JanStaschulat committed:
Merge branch 'feature/merge-two-pages-about-client-libraries' of https://github.com/boschresearch/ros2_documentation into feature/merge-two-pages-about-client-libraries

JanStaschulat commented:
> Besides the small change inline, we should also add in a redirect from the old (deleted) page to the new one. So at the top of `About-ROS-2-Client-Libraries`, add in:
> 
> ```
> .. redirect-from::
> 
>    Concepts/About-Client-Interfaces
> ```
> 
> (see https://github.com/ros2/ros2_documentation/blob/rolling/source/Concepts/About-Logging.rst?plain=1 for an example)

Sure, resolved in [858d0da](https://github.com/ros2/ros2_documentation/pull/2919/commits/858d0daed4d54bbca30e0bdc2ac78c4add178d66).

JanStaschulat committed:
removed two empty lines.

Signed-off-by: Jan Staschulat (CR/ADA1.2) <jan.staschulat@de.bosch.com>

clalancette reviewed (changes requested):
The change looks generally good to me, but when I build the documentation locally with this in place I get a bunch of warnings.

One of the warnings is that `Concepts/About-Client-Interfaces` is still referenced from https://github.com/ros2/ros2_documentation/blob/4b4c32c44d306f037a70b92e53b7b42567ce430c/source/Concepts.rst?plain=1#L37 , so that needs to be fixed.

The others all revolve around "Undefined substitution reference".  I think that you'll need to add `.. include:: ../../global_substitutions.txt` near the top of `About-ROS-2-Client-Libraries` for this to work.

fujitatomoya reviewed (approved):
lgtm with https://github.com/ros2/ros2_documentation/pull/2919#pullrequestreview-1065486398

JanStaschulat committed:
added '.. include:: ../../global_substitutions.txt' changed redirect to 'About-Client-Interfaces' without suffix 'Concepts' like in 'About-RQt.rst', maybe this works better.

Signed-off-by: Jan Staschulat (CR/ADA1.2) <jan.staschulat@de.bosch.com>

JanStaschulat commented:
@clalancette thanks for your great feedback. I've done the following:
- changed redirect name to `About-Client-Interfaces` without the folder name `Concepts/`. Maybe this works better? I saw such redirects to a file, which is located in the same folder, in  [About-RQt.rst](https://github.com/ros2/ros2_documentation/blob/rolling/source/Concepts/About-RQt.rst)
- replaced the link to `About-Client-Interfaces` by  `About-ROS-2-Client-Libraries` in [`source/Concepts.rst`](https://github.com/ros2/ros2_documentation/pull/2919/files)
- added include statement for global substitutions.

Does it build without warnings now?

clalancette commented on source/Concepts/About-ROS-2-Client-Libraries.rst:
```diff
@@ -1,3 +1,9 @@
+.. redirect-from::
+
+   About-Client-Interfaces
```
```suggestion
   Concepts/About-Client-Interfaces
```

(this has to be the path relative to the top of the site)

clalancette reviewed (changes requested):
One last thing to fix, then this will be good to go.

JanStaschulat committed:
Update source/Concepts/About-ROS-2-Client-Libraries.rst

Co-authored-by: Chris Lalancette <clalancette@gmail.com>

JanStaschulat committed:
added section heading

Signed-off-by: Jan Staschulat (CR/ADA1.2) <jan.staschulat@de.bosch.com>

JanStaschulat commented:
local `make test` is ok: 

```
jst3si@RNGX5596:~/github/src/ros2_documentation$ make test
doc8 --ignore D001 --ignore-path build
Scanning...
Validating...
========
Total files scanned = 228
Total files ignored = 213
Total accumulated errors = 0
Detailed error counts:
    - doc8.checks.CheckCarriageReturn = 0
    - doc8.checks.CheckIndentationNoTab = 0
    - doc8.checks.CheckMaxLineLength = 0
    - doc8.checks.CheckNewlineEndOfFile = 0
    - doc8.checks.CheckTrailingWhitespace = 0
    - doc8.checks.CheckValidity = 0
```
local `make linkcheck` is ok:

```
jst3si@RNGX5596:~/docu$ grep About-ROS-2-Client-Libraries ros2_docu_links_check.txt 
    Update source/Concepts/About-ROS-2-Client-Libraries.rst
reading sources... [100%] Concepts/About-ROS-2-Client-Libraries                 
reading sources... [100%] Concepts/About-ROS-2-Client-Libraries                 
(Concepts/About-ROS-2-Client-Libraries: line   52) ok        http://docs.ros.org/en/rolling/p/rclcpp
(Concepts/About-ROS-2-Client-Libraries: line   69) redirect  http://docs.ros2.org/foxy/api/rclpy/index.html - permanently to https://docs.ros2.org/foxy/api/rclpy/index.html
(Concepts/About-ROS-2-Client-Libraries: line  104) ok        https://docs.ros2.org/latest/api/rcl/
(Concepts/About-ROS-2-Client-Libraries: line   77) ok        https://github.com/ada-ros/ada4ros2
(Concepts/About-ROS-2-Client-Libraries: line   79) ok        https://github.com/ros2-java
(Concepts/About-ROS-2-Client-Libraries: line   82) ok        https://github.com/ros2-rust/ros2_rust
(Concepts/About-ROS-2-Client-Libraries: line   78) ok        https://github.com/ros2/rclc
(Concepts/About-ROS-2-Client-Libraries: line   49) ok        https://github.com/ros2/rclcpp
(Concepts/About-ROS-2-Client-Libraries: line   66) ok        https://github.com/ros2/rclpy
(Concepts/About-ROS-2-Client-Libraries: line   78) ok        https://micro.ros.org/
(Concepts/About-ROS-2-Client-Libraries: line  115) ok        https://roscon.ros.org/2016/presentations/ROSCon%202016%20-%20ROS%202%20Update.pdf
(Concepts/About-ROS-2-Client-Libraries: line  115) ok        https://vimeo.com/187696091
(Concepts/About-ROS-2-Client-Libraries: line   81) ok        https://www.npmjs.com/package/rclnodejs
(Concepts/About-ROS-2-Client-Libraries: line   87) ok        https://github.com/esteve/ros2_objc
(Concepts/About-ROS-2-Client-Libraries: line   80) redirect  https://github.com/esteve/ros2_dotnet - permanently to https://github.com/ros2-dotnet/ros2_dotnet
(Concepts/About-ROS-2-Client-Libraries: line   86) ok        https://github.com/firesurfer/rclcs
jst3si@RNGX5596:~/docu$ 
```

clalancette reviewed (approved):
This looks fantastic, thanks!
